src/sorting_network.cpp

```diff
@@ -0,0 +1,199 @@
+  {
+    return unsorted;
+  }
+  else if (len == 2)
+  {
+    return sort_two(unsorted[0], unsorted[1]);
+  }
+
+  size_t pivot = len / 2;
+  auto begin = unsorted.begin();
+  TermVec left_vec(begin, begin + pivot);
+  TermVec right_vec(begin + pivot, unsorted.end());
+
+  TermVec left_res = sorting_network_rec(left_vec);
+  TermVec right_res = sorting_network_rec(right_vec);
+
+  return merge(left_res, right_res);
+}
+
+TermVec SortingNetwork::sort_two(const Term & t1, const Term & t2) const
+{
+  return { solver_->make_term(Or, t1, t2), solver_->make_term(And, t1, t2) };
+}
+
+TermVec SortingNetwork::merge(const TermVec & sorted1,
+                              const TermVec & sorted2) const
+{
+  size_t len1 = sorted1.size();
+  size_t len2 = sorted2.size();
+
+  if (len1 == 0)
+  {
+    return sorted2;
+  }
+  else if (len2 == 0)
+  {
+    return sorted1;
+  }
+  else if (len1 == 1 && len2 == 1)
+  {
+    return sort_two(sorted1[0], sorted2[0]);
+  }
+
+  bool sorted1_is_even = (len1 % 2) == 0;
+  bool sorted2_is_even = (len2 % 2) == 0;
+
+  if (!sorted1_is_even and sorted2_is_even)
+  {
+    // normalize so we don't have to do all cases
+    return merge(sorted2, sorted1);
+  }
+
+  TermVec even_sorted1;
+  TermVec odd_sorted1;
+
+  for (size_t i = 0; i < len1; ++i)
+  {
+    if (i % 2 == 0)
+    {
+      odd_sorted1.push_back(sorted1[i]);
+    }
+    else
+    {
+      even_sorted1.push_back(sorted1[i]);
+    }
+  }
+
+  TermVec even_sorted2;
+  TermVec odd_sorted2;
+
+  for (size_t i = 0; i < len2; ++i)
+  {
+    if (i % 2 == 0)
+    {
+      odd_sorted2.push_back(sorted2[i]);
+    }
+    else
+    {
+      even_sorted2.push_back(sorted2[i]);
+    }
+  }
+
+  TermVec res_even = merge(even_sorted1, even_sorted2);
+  TermVec res_odd = merge(odd_sorted1, odd_sorted2);
+  size_t len_res_even = res_even.size();
+  size_t len_res_odd = res_odd.size();
+
+  TermVec res;
+  res.reserve(len1 + len2);
+
+  if (sorted1_is_even && sorted2_is_even)
+  {
+    Term first_res_odd = res_odd[0];
+    res.push_back(first_res_odd);
+    for (size_t i = 0; i < len_res_odd - 1; ++i)
+    {
+      const TermVec & two_sorted = sort_two(res_odd[i + 1], res_even[i]);
+      assert(two_sorted.size() == 2);
+      res.push_back(two_sorted[0]);
+      res.push_back(two_sorted[1]);
+    }
+
+    Term last_res_even = res_even.back();
+    res.push_back(last_res_even);
+  }
+  else if (sorted1_is_even && !sorted2_is_even)
+  {
+    Term first_res_odd = res_odd[0];
+    res.push_back(first_res_odd);
+    for (size_t i = 0; i < len_res_even; ++i)
+    {
+      const TermVec & two_sorted = sort_two(res_odd[i + 1], res_even[i]);
+      assert(two_sorted.size() == 2);
+      res.push_back(two_sorted[0]);
+      res.push_back(two_sorted[1]);
+    }
+  }
+  else if (!sorted1_is_even && !sorted2_is_even)
+  {
+    Term first_res_odd = res_odd[0];
+    Term last_res_odd = res_odd.back();
+
+    res.push_back(first_res_odd);
+    for (size_t i = 0; i < len_res_even; ++i)
```

Comment:
Can you add documentation about these cases?

Reply:
Not the best documentation, but a little better done with d6bc10dfc4f2f1fba66463bb0187347524396e1f.